packages/sync-actions/src/product-types-actions.js

```diff
@@ -256,11 +256,56 @@ const generateUpdateActionsForAttributeEnumValues = (
   ]
 }
 
-export const actionsMapForHints = (nestedValuesChanges = {}) => [
-  ...generateUpdateActionsForAttributeDefinitions(
-    nestedValuesChanges.attributeDefinitions
-  ),
-  ...generateUpdateActionsForAttributeEnumValues(
-    nestedValuesChanges.attributeEnumValues
-  ),
-]
+const generateChangeAttributeOrderAction = (
+  attrsOld = [],
+  attrsNew = [],
+  updateActions = []
+) => {
+  if (!attrsOld.length || !attrsNew.length) return null
+
+  const orderNew = attrsNew.map(attribute => attribute.name)
+
+  const removedAttributeNames = updateActions
+    .filter(action => action.action === 'removeAttributeDefinition')
+    .map(action => action.name)
+
+  const addedAttributeNames = updateActions
+    .filter(action => action.action === 'addAttributeDefinition')
+    .map(action => action.attribute.name)
+
+  // changeAttributeOrder action will be sent to CTP API as the last action so we have to
+  // calculate how the productType will look like after adding/removing attributes
+  const patchedOrderOld = attrsOld
+    .map(attribute => attribute.name)
+    .filter(name => !removedAttributeNames.includes(name))
+    .concat(addedAttributeNames)
+
+  if (orderNew.join(',') !== patchedOrderOld.join(','))
```

Comment:
sorry, but isn't this the same as `orderNew.join() !=== attrsOld.map(a => a.name).join()`?

Reply:
Hi @adnasa .. thx for the comment. Let's assume, that user provided new attributes:
```
attr1, attr2, attr4
```
And the productType saved on API has following attributes in this order:
```
attr1, attr2, attr3
```
The productType sync would generate actions `remove attr3, add attr4` so the order which we should compare is not `attrsOld.map(a => a.name).join()` (alias: `attr1, attr2, attr3`) but patched productType with the new order `attr1, attr2, attr4` (because API will also perform actions for adding and removing attributes and then it will compare order).

We have to do this check because if we send `changeOrder` actions with the same order attributes currently have, API will throw an error.

Reply:
Thanks for clarifying. This sounds like a good candidate for a code comment ☝🏽
because there is no way I will remember this in the future